src/commands/force/source/open.ts

```diff
@@ -0,0 +1,160 @@
+/*
+ * Copyright (c) 2021, salesforce.com, inc.
+ * All rights reserved.
+ * Licensed under the BSD 3-Clause license.
+ * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+
+import * as os from 'os';
+import * as path from 'path';
+import * as open from 'open';
+import { getBoolean, getString } from '@salesforce/ts-types';
+import { fs, AuthInfo } from '@salesforce/core';
+import { SfdxCommand, flags, FlagsConfig } from '@salesforce/command';
+import { Messages, sfdc, SfdxError } from '@salesforce/core';
+import checkLightningDomain from '@salesforce/core/lib/util/checkLightningDomain';
+import { SourceComponent, MetadataResolver } from '@salesforce/source-deploy-retrieve';
+import { OpenResultFormatter, OpenCommandResult } from '../../../formatters/openResultFormatter';
+
+export interface DnsLookupObject {
+  address: string;
+  family: number;
+}
+
+export interface FlexiPageRecord {
+  attributes: {
+    type: string;
+    url: string;
+  };
+  Id: string;
+}
+
+Messages.importMessagesDirectory(__dirname);
+const messages = Messages.loadMessages('@salesforce/plugin-source', 'open');
+
+export class Open extends SfdxCommand {
+  public static readonly description = messages.getMessage('description');
+  public static readonly examples = messages.getMessage('examples').split(os.EOL);
+  public static readonly requiresProject = true;
+  public static readonly requiresUsername = true;
+  public static readonly flagsConfig: FlagsConfig = {
+    sourcefile: flags.filepath({
+      char: 'f',
+      required: true,
+      description: messages.getMessage('SourceOpenFileDescription'),
+    }),
+    urlonly: flags.boolean({
+      char: 'r',
+      description: messages.getMessage('SourceOpenPathDescription'),
+    }),
+  };
+
+  protected openResult: OpenCommandResult;
+
+  public async run(): Promise<OpenCommandResult> {
+    await this.doOpen();
+    this.resolveSuccess();
+    return this.formatResult();
+  }
+
+  private async doOpen(): Promise<void> {
+    const typeName = this.getTypeNameDefinitionByFileName(path.resolve(this.flags.sourcefile));
+    const openPath = typeName === 'FlexiPage' ? await this.handleSupportedTypes() : await this.handleUnsupportedTypes();
+
+    this.openResult = await this.open(openPath);
+  }
+
+  private getTypeNameDefinitionByFileName(fsPath: string): string | undefined {
+    if (fs.fileExistsSync(fsPath)) {
+      const metadataResolver = new MetadataResolver();
+      const components: SourceComponent[] = metadataResolver.getComponentsFromPath(fsPath);
+      return components[0].type.name;
```

Comment:
is there guaranteed to be only 1 result here?

Reply:
`getComponentsFromPath` always returns an array even for directories, we pick the first component type out of that list, ideally user should pick a single `flexipage`.
If file/directory exist but cannot determine source type the method throws an error `Could not infer a metadata type`